src/api/flow.ts

```diff
@@ -117,53 +117,58 @@ export function createFlowGenerator(name: string, generator: Function) {
         const args = arguments
         const runId = ++generatorId
         const gen = action(`${name} - runid: ${runId} - init`, generator).apply(ctx, args)
-        let stepId = 0
-        let resolver: (value: any) => void
         let rejector: (error: any) => void
 
-        function onFulfilled(res: any) {
-            let ret
-            try {
-                ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.next).call(
-                    gen,
-                    res
-                )
-            } catch (e) {
-                return rejector(e)
+        const res = new Promise(function(resolve, reject) {
+            let stepId = 0
+            rejector = reject
+
+            function onFulfilled(res: any) {
+                let ret
+                try {
+                    ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.next).call(
+                        gen,
+                        res
+                    )
+                } catch (e) {
+                    return reject(e)
+                }
+                next(ret)
+                return null
             }
-            next(ret)
-            return null
-        }
 
-        function onRejected(err: any) {
-            let ret
-            try {
-                ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.throw).call(
-                    gen,
-                    err
-                )
-            } catch (e) {
-                return rejector(e)
+            function onRejected(err: any) {
+                let ret
+                try {
+                    ret = action(`${name} - runid: ${runId} - yield ${stepId++}`, gen.throw).call(
+                        gen,
+                        err
+                    )
+                } catch (e) {
+                    return reject(e)
+                }
+                next(ret)
             }
-            next(ret)
-        }
 
-        function next(ret: any) {
-            if (ret.done) return resolver(ret.value)
-            // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
-            if (!ret.value || typeof ret.value.then !== "function")
-                return fail("Only promises can be yielded to asyncAction, got: " + ret)
-            return ret.value.then(onFulfilled, onRejected)
-        }
+            function next(ret: any) {
+                if (ret.done) return resolve(ret.value)
+                // TODO: support more type of values? See https://github.com/tj/co/blob/249bbdc72da24ae44076afd716349d2089b31c4c/index.js#L100
+                if (!ret.value || typeof ret.value.then !== "function")
```

Comment:
Native async functions allow `await`ing arbitrary things, although I still think we shouldn't roll our own coroutines if we don't also get async generators just for cancellation.